azuredevops/internal/service/memberentitlementmanagement/resource_service_principal_entitlement.go

```diff
@@ -0,0 +1,356 @@
+	}
+	licensingSource, ok := d.GetOk("licensing_source")
+	if !ok {
+		return fmt.Errorf(" Reading account licensing source for ServicePrincipalEntitlementID: %s", servicePrincipalEntitlementID)
+	}
+
+	clients := m.(*client.AggregatedClient)
+
+	patchResponse, err := clients.MemberEntitleManagementClient.UpdateServicePrincipalEntitlement(clients.Ctx,
+		memberentitlementmanagement.UpdateServicePrincipalEntitlementArgs{
+			ServicePrincipalId: &id,
+			Document: &[]webapi.JsonPatchOperation{
+				{
+					Op:   &webapi.OperationValues.Replace,
+					From: nil,
+					Path: converter.String("/accessLevel"),
+					Value: struct {
+						AccountLicenseType string `json:"accountLicenseType"`
+						LicensingSource    string `json:"licensingSource"`
+					}{
+						string(*accountLicenseType),
+						licensingSource.(string),
+					},
+				},
+			},
+		})
+
+	if err != nil {
+		return fmt.Errorf(" Updating service principal entitlement: %v", err)
+	}
+
+	if !*patchResponse.IsSuccess {
+		return fmt.Errorf(" Updating service principal entitlement: %s", getServicePrincipalEntitlementAPIErrorMessage(patchResponse.OperationResults))
+	}
+	return resourceServicePrincipalEntitlementRead(d, m)
+}
+
+func resourceServicePrincipalEntitlementDelete(d *schema.ResourceData, m interface{}) error {
+	if d.Id() == "" {
+		return nil
+	}
+
+	servicePrincipalEntitlementID := d.Id()
+	id, err := uuid.Parse(servicePrincipalEntitlementID)
+	if err != nil {
+		return fmt.Errorf(" Parsing ServicePrincipalEntitlement ID. ServicePrincipalEntitlementID: %s. %v", servicePrincipalEntitlementID, err)
+	}
+
+	clients := m.(*client.AggregatedClient)
+
+	err = clients.MemberEntitleManagementClient.DeleteServicePrincipalEntitlement(m.(*client.AggregatedClient).Ctx, memberentitlementmanagement.DeleteServicePrincipalEntitlementArgs{
+		ServicePrincipalId: &id,
+	})
+
+	if err != nil {
+		return fmt.Errorf(" Deleting service principal entitlement: %v", err)
+	}
+
+	return nil
+}
+
+func expandServicePrincipalEntitlement(d *schema.ResourceData) (*memberentitlementmanagement.ServicePrincipalEntitlement, error) {
+	origin := d.Get("origin").(string)
+	originID := d.Get("origin_id").(string)
+	displayName := d.Get("display_name").(string)
+	descriptor := d.Get("descriptor").(string)
+
+	accountLicenseType, err := converter.AccountLicenseType(d.Get("account_license_type").(string))
+	if err != nil {
+		return nil, err
+	}
+	licensingSource, err := converter.AccountLicensingSource(d.Get("licensing_source").(string))
+	if err != nil {
+		return nil, err
+	}
+
+	return &memberentitlementmanagement.ServicePrincipalEntitlement{
+		AccessLevel: &licensing.AccessLevel{
+			AccountLicenseType: accountLicenseType,
+			LicensingSource:    licensingSource,
+		},
+		ServicePrincipal: &graph.GraphServicePrincipal{
+			Origin:      &origin,
+			OriginId:    &originID,
+			DisplayName: &displayName,
+			Descriptor:  &descriptor,
+			SubjectKind: converter.String("servicePrincipal"),
+		},
+	}, nil
```

Comment:
Since the `AccountLicenseType` and `LicensingSource`  has been constrained in the scheme definition, we can directly  use it without checking the values
```suggestion
	return &memberentitlementmanagement.ServicePrincipalEntitlement{
		AccessLevel: &licensing.AccessLevel{
			AccountLicenseType: converter.ToPtr(licensing.AccountLicenseType(d.Get("account_license_type").(string))),
			LicensingSource:    converter.ToPtr(licensing.LicensingSource(d.Get("licensing_source").(string))),
		},
		ServicePrincipal: &graph.GraphServicePrincipal{
			Origin:      converter.ToPtr(d.Get("origin").(string)),
			OriginId:    converter.ToPtr(d.Get("origin_id").(string)),
			DisplayName: converter.ToPtr(d.Get("display_name").(string)),
			Descriptor:  converter.ToPtr(d.Get("descriptor").(string)),
			SubjectKind: converter.String("servicePrincipal"),
		},
	}, nil
```